vllm/model_executor/layers/rotary_embedding.py

```diff
@@ -232,6 +232,22 @@ def forward_tpu(
                       if self.use_native2 else self.forward_native)
         return forward_fn(positions, query, key, offsets)
 
+    def forward_cpu(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        offsets: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Note: the forward_native() with torch.compile has significant
+        # performance regression.
```

Comment:
QQ: Have you tried `forward_native2` which was added in the TPU integration PR? For TPUs, `forward_native2` works much better than `forward_native`.

Reply:
I did a quick test but it still has significant performance regression. I also saw a warning that torchInductor doesn't support complex operations (```freqs_cis``` is complex type).